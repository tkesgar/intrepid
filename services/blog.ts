import path from "path";
import fs from "fs/promises";
import yaml from "yaml";

interface Post {
  content: string;
  title: string;
  slug: string;
  date: Date;
  hidden: boolean;
}

function toTitleCase(str: string) {
  return str.replaceAll(/\w\S*/g, (substr) => {
    const head = substr.slice(0, 1);
    const tail = substr.slice(1);
    return [head.toUpperCase(), tail.toLowerCase()].join("");
  });
}

export async function getPost(
  dirName: string,
  opts: {
    blogPath?: string;
    timeOffset?: string;
  } = {}
): Promise<Post> {
  const { blogPath = "./blog", timeOffset = "+07:00" } = opts;

  const postPath = path.join(process.cwd(), blogPath, dirName);

  const [content, meta] = await Promise.all([
    (async () => {
      const contentPath = path.join(postPath, "content.md");
      const content = await fs.readFile(contentPath, { encoding: "utf-8" });

      return content;
    })(),
    (async () => {
      const metaPath = path.join(postPath, "meta.yml");

      let metaYaml: string;
      try {
        metaYaml = await fs.readFile(metaPath, { encoding: "utf-8" });
      } catch (error) {
        // If we cannot read the meta (the file does not exist), fall back to
        // a default meta retrieved from directory name.
        const [strDate, strTitle] = dirName.split(".");

        return {
          title: toTitleCase(strTitle.replaceAll(/[-]/g, " ")),
          slug: strTitle,
          date: new Date(`${strDate}T00:00:00${timeOffset}`),
          hidden: false,
        };
      }

      // Splitting parsing the YAML here so that if it errors (e.g. invalid syntax),
      // the error will propagate out.
      const meta = yaml.parse(metaYaml);

      return {
        title: meta.title,
        slug: meta.slug,
        date: new Date(meta.date),
        hidden: meta.hidden,
      };
    })(),
  ]);

  return {
    content,
    title: meta.title,
    slug: meta.slug,
    date: meta.date,
    hidden: meta.hidden,
  };
}
