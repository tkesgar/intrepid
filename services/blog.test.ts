import { getPost } from "./blog";

const TEST_BLOG_PATH = "./tests/test-blog";

describe("getPost", () => {
  it("should return the post data", async () => {
    const post = await getPost("2021-07-08.test", { blogPath: TEST_BLOG_PATH });

    expect(post).toMatchSnapshot();
  });

  it("should return the complete meta data from meta.yml", async () => {
    const post = await getPost("2021-07-08.test", { blogPath: TEST_BLOG_PATH });

    expect(post.title).toBe("Test Post");
    expect(post.slug).toBe("test-post");
    expect(post.date).toEqual(new Date("2021-07-08T21:56:39+07:00"));
  });

  it("should return the post data without meta.yml", async () => {
    const post = await getPost("2021-07-08.test-no-meta", {
      blogPath: TEST_BLOG_PATH,
    });

    expect(post.title).toBe("Test No Meta");
    expect(post.slug).toBe("test-no-meta");
    expect(post.date).toEqual(new Date("2021-07-08T00:00:00+07:00"));
  });
});
