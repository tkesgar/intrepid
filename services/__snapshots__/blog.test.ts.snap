// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPost should return the post data 1`] = `
Object {
  "content": "# Hello world!

It works!
",
  "date": 2021-07-08T14:56:39.000Z,
  "hidden": true,
  "slug": "test-post",
  "title": "Test Post",
}
`;
